core/state/trie_prefetcher.go

```diff
@@ -251,42 +242,30 @@ func newSubfetcher(db Database, state common.Hash, owner common.Hash, root commo
 func (sf *subfetcher) schedule(keys [][]byte) {
 	// Append the tasks to the current queue
 	sf.lock.Lock()
+
 	sf.tasks = append(sf.tasks, keys...)
 	sf.lock.Unlock()
-
-	// Notify the prefetcher, it's fine if it's already terminated
+	// Notify the prefetcher. The wake-chan is buffered, so this is async.
 	select {
-	case sf.wake <- struct{}{}:
+	case sf.wake <- true:
 	default:
 	}
 }
 
-// peek tries to retrieve a deep copy of the fetcher's trie in whatever form it
-// is currently.
-func (sf *subfetcher) peek() Trie {
-	ch := make(chan Trie)
-	select {
-	case sf.copy <- ch:
-		// Subfetcher still alive, return copy from it
-		return <-ch
-
-	case <-sf.term:
-		// Subfetcher already terminated, return a copy directly
-		if sf.trie == nil {
-			return nil
-		}
-		return sf.db.CopyTrie(sf.trie)
-	}
-}
-
-// abort interrupts the subfetcher immediately. It is safe to call abort multiple
+// wait waits for the subfetcher to finish it's task. It is safe to call wait multiple
 // times but it is not thread safe.
-func (sf *subfetcher) abort() {
-	select {
-	case <-sf.stop:
-	default:
-		close(sf.stop)
+func (sf *subfetcher) wait() {
+	// Signal termination by nil tasks
+	sf.lock.Lock()
+	if sf.closing {
+		sf.lock.Unlock()
+		return // already exiting
```

Comment:
This will not `wait`, rather return early, even if shutdown is still being processed (and someone else requested wait already).

Reply:
It says in the comment description that this function is not thread safe.